### PR TITLE
DAOS-330 pool: Consider FDs in PS placement

### PR DESCRIPTION
**Pull request by liw**
Currently, when placing PS replicas, we may unknowingly choose multiple
engines on the same server while all engines are vacant on other
servers, because we do not look at FDs in the process. This patch
improves PS placement to consider server-level FDs and adds some
randomness to choices:

  - Use ds_pool_plan_svc_reconfs when creating pools too.

  - Consider server-level FDs in ds_pool_plan_svc_reconfs, so that a
    domain gets i replicas only if all domains that can host fewer than
    i replicas are full.

  - Shuffle the list of domains we consider in ds_pool_plan_svc_reconfs,
    and choose engines within a domain randomly.

  - Skip the later steps in ds_pool_plan_svc_reconfs when planning
    "synchronous removals" right before updating pool maps, for we
    should do only the removals strictly required by each pool map
    update---the "later step" may introduce removals for svc_rf and/or
    balancing.

  - Fix some tests that break when PS placement algorithm changes.

  - Always sort PS rank lists, so that command outputs and log messages
    will be easier to work with.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by liw** [comment]
The previous push has passed the CI testing; the current push improves comments and commit message.

**Review by kccain — APPROVED** [review]
only a couple of very minor typo suggestions
Still need to go through the test functions in srv_util.c, will add any new comments as single comments in that area.

**Comment by kccain** [comment]
I've concluded my review now.

**Review by liw — COMMENTED** [review]
@kccain, thanks for the careful review.

**Review by kccain — APPROVED** [review]

**Comment by liw** [comment]
The test failures are due to https://daosio.atlassian.net/browse/DAOS-14598.

**Review by liuxuezhao — APPROVED** [review]

File: src/pool/srv_util.c [hunk 1]
```diff
@@ -220,146 +220,451 @@ ds_pool_transfer_map_buf(struct pool_buf *map_buf, uint32_t map_version,
...
+		struct reconf_map *rmap_out, d_rank_list_t *to_remove)
+{
+	struct reconf_map   rmap = {0};
+	struct pool_domain *domains;
+	int                 domains_len;
+	d_rank_list_t      *replicas_left = NULL;
+	int                 i;
+	int                 rc;
+
+	domains_len = pool_map_find_domain(map, PO_COMP_TP_NODE, PO_COMP_ID_ALL, &domains);
```
[the remark this answers is not in the document]
Comment (liw) [reply src/pool/srv_util.c:304]: I have verified with Kris: "Hi Kris. Is it true that on master we always have PO_COMP_TP_NODE level domains in pool maps?" "Correct, it is always there". If this changes in the future, I'll have to simulate such a domain level...

**Comment by liw** [comment]
Ping @daos-stack/daos-gatekeeper...

**Pull request merged**